### DISCUSSION TITLE
Web Interface Stack Upgrade

### DISCUSSION THREAD

**Discussion by sbalbrecht** [sabnzbd/sabnzbd#2777] [Ideas]
Following up from the beginning of [this discussion](https://github.com/sabnzbd/sabnzbd/pull/2774#issuecomment-1885811404). I am new to SABnzbd and open source contributing but I saw a need here and feel motivated to make some contributions.

The technologies behind SABnzbd’s UI haven't been updated in some years, with some being leftover from even earlier refactors. Many of the libraries used are old and/or unmaintained, and some obsolete. This discussion is to analyze and understand how the current stack is used and form a plan to migrate to more modern and maintainable alternatives.

The goal is for the change to be transparent to the end-user. Changing the look and feel of the UI is not part of this upgrade plan, but could be a separately-planned project.

I started writing a section on potential UI frameworks but there is a lot to unpack, so I thought it would be wise to split that into a separate discussion after establishing a high-level plan.

## Current Stack

These are the libraries involved in the UI from what I've seen poking around the project.

- Cheetah
    - All HTML templates are defined with Cheetah and its control flow is heavily used in templates to generate the UI
    - It is an old, small project, so it would be desirable to migrate to a modern UI framework (e.g., Angular, React, Vue, Svelte).
    - Still poking around, does it use server-side rendering or are the files statically generated?
- jQuery
    - JavaScript is now capable of doing everything jQuery does, albeit less concisely.
    - Used heavily in the project for conveniently selecting elements and maybe making AJAX requests.
- Knockout.js
    - Used to bind Javascript variables to the HTML, and update the UI when the state of the JS changes. MVVM pattern.
    - Hadn’t heard of this library before, but its functionality is a basic feature of any modern UI framework
- moment.js
    - Super useful in its day, but is now in maintenance mode. It is a heavy dependency for what it provides.
    - The project itself recommends [switching to more modern libraries](https://momentjs.com/docs/#/-project-status/) like Luxon or day.js (much lighter and actively maintained)
- Bootstrap CSS components
    - The bootstrap look and feel has been heavily modified so it seems mainly used as utility and convenience (quickly styling buttons, a working navbar, organizing elements in a grid).
    - The version of Bootstrap in the project still requires jQuery, while new versions do not. If we needed to stick with Bootstrap we could still eliminate jQuery.
    - CSS has improved a lot since Bootstrap’s inception. CSS grid and a load of other convenience features make a lot of Bootstrap’s utility less valuable.
- Any others I’m missing?

## Potential Migrations

1. jQuery can likely be eliminated without insane effort. jQuery is used heavily but the migration steps to native JS are [fairly straightforward](https://tobiasahlin.com/blog/move-from-jquery-to-vanilla-javascript/). Eliminating jQuery (besides its required use in Bootstrap) would probably be a good first step.
2. Moment.js does not appear to be used for too much. I haven’t looked into how the locales would migrate to another library like Luxon/day.js but I would be surprised if this was a big issue. Eliminating moment would be a good second step, but more research would have to be done to decide on the best replacement. I haven't researched Luxon but I know Day.js is tiny, designed to have a similar API to moment.
3. Conceptually I am lumping together a migration from both Cheetah and Knockout.js. The functionality of both projects should translate easily to a modern UI framework, both in templating with reusable components and binding and updating the UI in response to changes in state. We can review possible candidates separate from this discussion. This would be a good third step.
4. We could migrate from Bootstrap to either native CSS or something like Tailwind, or simply upgrade Bootstrap to the latest version. It depends on how and why it is used in the project, so I’d like more context on that. With modern UI frameworks making it easy to build isolated components, I think the meta has shifted more heavily in favor of Tailwind’s approach. The main opposition I’ve seen to tailwind is that it can bloat the HTML with many classes, but I’ve found with components this is less of a concern. [Design philosophy behind Tailwind](https://adamwathan.me/css-utility-classes-and-separation-of-concerns/).

**Comment by Safihre** [comment]
Good analysis and I'd like to jump in straight away on your suggestions for potential migrations:
As you are a software developer you probably agree with me (I'm a product manger at a software company)  that we should try to keep the stories/changes as small as possible and try to prevent waterfall deliveries.
The UI stack for Glitter and the Config are slightly different, but they do share a few common components like Cheetah (server side template rendering) and the API. We should probably seperate changes to the Config and Glitter to keep things as light as possible.
Your first suggestion is jQuery, but I can already tell you that unfortunately it's used much more than appears from first glance. There's of course the Bootstrap usage, but we also use jQuery-UI for the drag-and-drop, and several plugfins like typeahead.
So, I would not start with that one, I would suggest the following:

1. Replacing Cheetah by modern template alternative. This way you can already see large parts of the UI code, identify parts that can be replaced by API-calls and others that _should_ be replaced by API calls instead of template-based logic.
2. Update the API to use a modern engine to add new API-calls in a more modern way. @jcfp or I can do that.
3. Update Bootstrap and jQuery.
4. Update UI-library.
The last ones are deliberately unspecific, as we will have to see how well the first steps go 😁

What do you think?

**Reply by sbalbrecht** [reply]
> we should try to keep the stories/changes as small as possible and try to prevent waterfall deliveries

100% agreed on that 😁

Point taken on jQuery. I assumed it was just replacing usages of the `$()` API but I can see how plugins, drag and drop would complicate eliminating it early. I'm fine with leaving that for later, as the framework will ease the replacement of those features.

I'm trying to wrap my head around migrating the config and glitter to a new framework separately. I assumed they share state and were considered part of the same application, which would make it difficult to migrate them separately. I guess I need clarification on how they interact. Certain frontend frameworks could make those isolated updates easier.

I also did not factor in that the API would also need updated but that makes sense. I lead an architecture team at work (Java) but I started my career in frontend. I haven't worked with Python much but I wouldn't shy away from giving input on the API. How will Starlette change how API calls are made compared to Cherrypy?

> identify parts that can be replaced by API-calls and others that should be replaced by API calls instead of template-based logic.

Do you have any examples of this off the top of your head? Just curious to have a frame of reference, but I am comfortable with this idea and the order of tasks.

**Reply by Safihre** [reply]
Aah that's good to know that you are fresh to Python and strong in front end. Of course it's not hard language at all, but then it's probably better that you start the same way that I started many years ago: replacing the front end with newer techniques and when you run into something that the API can't do, you ask me to add it :)

Replacing Cherrypy actually won't change the API, it's just something that is related to this but can be done independently and has been on the roadmap for a while so it's on my mind. You can ignore that completely for now.

 You can reach out to me at safihre@sabnzbd.org and we can schedule an introduction call, if you really want to dive in 😃

**Reply by thezoggy** [reply]
i would assume the objective would move from a monolith to a microservices setup. 
where we can actually segment the frontend from the backend. 

about the backend stuff, specifically:
https://github.com/sabnzbd/sabnzbd/issues/2180#issuecomment-1128075368

to add on that, our api is kinda the fragile thing because its been around so long and so tied to the front end.. so makes it challenging at times to change stuff without possibly introducing regressions.  we do generate some test scenarios but of course its quite lacking to what ultimately we would need let alone validation of models/fuzzing input/etc. 

in this modern age we would define out api with openapi and have pydantic/flask-smorest w/marshmallow  (validation model) so we can validate+test out of the gate. 

then that also gets into the whole db stuff as sqlite is single threaded and can be a limiting factor for performance..  so switching to ORM might open the door to better optimized things which may handle larger scale setups (crazy huge queue/history?) 


but of course any of these pieces brings a bit of challenges and some are more important than others..

**Reply by Safihre** [reply]
Agreed, if this was a brand new product. However, knowing we'd probably have to keep the Legacy API (aka the current one) for a few more years, I think it's not a priority right now for such a big refactor.
The aging front end stack is worse and cherrypy needs to be removed.

**Reply by mnightingale** [reply]
Generally I think I agree with the order.

I'd favour React, Typescript, Vite, etc. but I won't get into that too much if @sbalbrecht wants to save it for another discussion. Sab would just serve the built SPA, node would run the web during development with vite proxying api calls to Sab.

I only mention that because trying to use jQuery with React or similar would be a really bad idea. Bootstrap would be less of a problem see react-bootstrap.

I agree the legacy API would likely have to remain for quite some time. For any new stuff I agree with @thezoggy, I would define an openapi specification and autogenerate the server and client with type checking.

I wondered when the issue discussion would evolve into discussing the web stack 😄

**Reply by sbalbrecht** [reply]
I'm fine with shooting some ideas around here, it can be formalized later. 

I've used Angular heavily in the past and while its opinionated structure is good for bigger teams/projects/maybe open source, the developer experience isn't fantastic due to all the ceremony. I also think since the other frameworks were built partially in response to its design flaws, at this point it's on its way out (even if it's still widely used).

I have not worked with React and Next.js or Vue and Nuxt.js so I can't speak to the developer experience, though I've heard that compared to Angular the project structure can vary widely since they don't box you in as much.

React is now by far the most popular framework and has only increased in popularity over the last few years, which gives it some major cred. Vue was a major competitor for a while but seems to have been decreasing in popularity over the last few years, but I can't speak much to it.

I have also worked with Svelte + SvelteKit. While relatively newer (2016, about to release highly anticipated v5 this year) and less widely used than the other frameworks, development is backed by Vercel and it has ranked consistently high as one of the most admired and loved frameworks for its developer experience. I'm certainly biased but I don't think it should be discarded based on its size.

For an open source project it would probably be wise to pick something widely used and popular like React, but its worth discussing.

> trying to use jQuery with React or similar would be a really bad idea

This was also my motivation to remove it before the framework migration. I worry it would be tough in any framework since they like to own their DOM manipulation.

**Reply by sbalbrecht** [reply]
[Some data so it's not all anecdotes](https://docs.google.com/spreadsheets/d/1kODyUrTPWvz5n0fpUovRdxfXsSikVHz6T3h9Kspuk8g/edit?pli=1#gid=81955775) 😉 
SO developer survey results from the last few years.

**Reply by sbalbrecht** [reply]
@Safihre and I chatted yesterday to help me get acquainted with the project and some of the caveats to this kind of migration. I'd like to follow up with some thoughts I've had since that discussion (turned out to be more thoughts than I anticipated… sorry).

I think we're feeling good about React -- it's super popular and accessible with a huge ecosystem. And honestly, if you wanted to migrate to a new framework in the future, it will probably be easier than this migration from Cheetah/Knockout/jQuery 😆

I largely agree with @mnightingale for the approach:

- The UI would likely be built with React and Typescript
    - @mnightingale: you mentioned Vite instead a framework like Next.js. Do you think Next.js is overkill? The React docs seem to [heavily push people towards it](https://react.dev/learn/start-a-new-react-project#can-i-use-react-without-a-framework) for its solutions to routing and data fetching, although we won’t use its server-side features. Without Next.js, how tough would it be set up a custom build process with Vite, and won't we still need to pull in libraries for routing & data fetching?
- It doesn’t make sense to run node.js in production alongside sab to leverage SSR. Node would only be used for a dev server during development and for building the static HTML/CSS/JS.
- Sab will simply serve the static web files that are compiled by node during the build stage

# Impact of introducing node, React

Before committing to adding a node-based project for the UI, I want to talk through the potential impact of this decision.

## Docker image

Node will not be required to run inside the sab Docker image -- the image can build the UI in a separate layer/image that can be discarded, leaving the build artifacts for use in the main image. The Dockerfile will become more complex but the container shouldn’t increase in size.

## Development

For development, there are now more steps to run the project. Launch UI dev server, launch SABnzbd.py. There is also more overhead to producing release artifacts. It’s not a ton but it is a little annoying.

You now have to deal with issues with the node ecosystem in addition to the python ecosystem. Issues running and installing npm packages, etc. We want to keep dependencies to a minimum but it’s a consideration.

## Users building from source

Users building from source now need to have node installed on the system. This is a big departure from the simple Python install, and I'm not sure how to avoid it.

@Safihre suggested checking the build artifacts into source control. I'm not fond of generated files in source control -- the commits become noiser and have potential for error if the wrong version of the artifacts are committed. 

But, Saf also suggested doing this only during release commits (since this consideration is for users building releases in source code), which sounds a lot more reasonable and predictable.

Or, just accept that users building from source will require an extra runtime and couple steps to build and run the project. It’s not the end of the world.

So if you want to minimally impact users building from source, I think including the latest web dist in the source as part of the release process is a fair tradeoff, but requires modifying any release automation logic. We would also need to maybe play with the .gitignore or other configurations since the same dist directory checked into VCS presumably will be used during development and its contents shouldn’t be updated by regular commits.

## Client-side Data Fetching

All the data that is currently inserted into the template server-side must now be fetched client-side. In most cases I don't anticipate high latency since most users probably access the UI on their LAN, but we will need to amend the client-side data access logic to include loading spinners/other feedback to indicate that data is being fetched from the server before we can completely render parts of the page. I'd like input from folks on how this could be mitigated.

## the good stuff

With those headaches out of the way, we would also gain a lot!

The UI will become a lot easier to work with, ideally through further componentization and leveraging modern QoL features of these frameworks. The developer experience with React and Next.js is clearly great or else it wouldn’t be as popular as it is. The UI will be on a platform that has a lot of momentum behind it and a large ecosystem of solutions that should make implementing new features easier.

# Closing thoughts

So, adding node has tradeoffs. I did some searching to see what things looked like if we wanted to keep everything in Python, and it wasn’t amazing. There are a lot of templating frameworks, some full-stack frameworks with templating support (i.e., Django), but it seems like you’d still need to supplement the client-side functionality with JS. I think it would be nice to keep the project Python-based for simplicity, but I also think going to node will make for a better UI development experience.

So… thoughts? 🤔 I know this is a lot to unpack, so thanks for the time.

Also, currently brainstorming how to migrate those `$T()` localization calls if the UI is all in node (they’ll no longer be filled into the template by the server). Query all translatable strings for a given page from the api on page load, then populate them? Would need a new localization endpoint.

**Reply by mnightingale** [reply]
I have worked on a few Next.js projects but I think it would be overkill for Sab and not a particularly good fit, there would be a lot of features unavailable because they rely on the server.

[create-react-app](https://create-react-app.dev) (CRA) used to be the simplest way to get started with React and was the documentations recommendation. I’m not well informed on the reasoning behind this change but I’d expect due to the limitations of performance, SEO and client-side rendering when many users wanted to use the stack to create public websites. As you're aware the recommendation is a framework more optimal for websites Next.js, Remix, Gatsby, etc. I don’t know enough about them to say if they’d be suitable for Sab static build requirements.

Vite isn’t a framework, it’s a build tool (CRA uses Webpack) focussing on a fast developer experience, quick startup, hot module replacement, no bundling in development. The Vite quick start template is similar to the CRA approach.

You might find it interesting to checking out the approach [autobrr/autobrr](https://github.com/autobrr/autobrr) takes, apart from the web part they template the index html to pull in some of the config options. The web part is quite a simple stack which relies on Vite, client routing with react-router and simplified data fetching and mutations with @tanstack/react-query.

---

**RE: Impact of introducing node**

**Docker image:** agree

**Development:** agree, the release workflow on Github would be quite simple; a job to build web and upload an artifact, building Sab would then just download that to package/include it in a release.

Users building from source: strongly agree on not committing build artifacts, can become very messy resolving conflicts. But it could be included in the source distribution artifact? Then users you want a source code distribution can just download that rather than cloning the repo?

Assuming users have node and whatever package manager is used compiling the frontend would only be a couple of commands.

**Client-side Data Fetching:** https://react.dev/reference/react/Suspense + Spinners I’d expect

**The good stuff:** I do think making the web aspects easier to work with and makes changes but be a huge improvement and certainly make it more feasible to make wholesale UI changes in the future, how it currently works is daunting and as has put me off attempting to make changes.

**Closing thoughts**

A few ways `$T()` could work, I’ve used https://react.i18next.com/latest/usetranslation-hook before loading multiple json files but it can also hit an api. There are tools for scanning to find used translations, so tooling could create lookup tables only for those which are used by the web..
Transifex do have their own packages, so you get the translations from their CDN but there are potentially costs or privacy concerns with that so probably best to package them in the project.

**Reply by sbalbrecht** [reply]
Thank you for this input!

Seeing autobrr’s setup, I think I agree that we should do without Next.js. It doesn’t look too complicated to set up, I just need to read up on Vite and some of those routing/data-fetching libs. SvelteKit actually uses Vite under the hood as well, but I never had to configure it since they abstract it away.

> But it could be included in the source distribution artifact? Then users you want a source code distribution can just download that rather than cloning the repo?

Ah, good call! That definitely sounds like the best route assuming we can modify what goes into that artifact.

Yeah, I see those Transifex packages for react that use their CDN… that’s too bad.

So after the user chooses their language, you’d just fetch their lang’s translation file from the API and hold it client-side, then the `useTranslation` hook just does its thing?

I have a lot to research now but I’m feeling comfortable with the plan. Will continue posting updates and questions here.

**Reply by sanderjo** [reply]
> Users building from source now need to have node installed on the system.

It's not only **building** from source, but also **running** from source, right? Or maybe that is the same?

Anyway: this might be a problem for users with an obscure setup. For example, I have SABnzbd running on Ubuntu 22.04.3 LTS on RISC-V. It looks like node.js is not avaible from Ubuntu's repo for RISC-V

```
sander@nezha:~$ sudo apt install nodejs
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Package nodejs is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source

E: Package 'nodejs' has no installation candidate
```
I browsed https://nodejs.org/en a bit, and I don't see a quick solution for that.

So for people with an obscure OS setup (RISC-V, NAS, Solaris), node.js might be a problem?

**Reply by Safihre** [reply]
Yeah that's actually why I want the compiled files to be part of the repository. Just like the Translations are auto updated by a Github actions we can do the same, that the developer doesn't commit the compiled files (due to being excluded through gitignore), but that the Github Actions force pushes them.

I know that's not the standard way, but us shipping basically a whole webserver for each installation isn't the usual setup.

**Reply by sbalbrecht** [reply]
Node is needed to compile the react code into static web files that sab could serve. If the pre-compiled files were packaged with release source code then you wouldn't need node unless you want to modify the UI code.

But yeah, looks like they're [working on supporting RISC-V](https://github.com/nodejs/build/issues/2876) but have some roadblocks.

Sab could maybe pull the compiled files from those releases if it detected the files were missing?

Edit: @Safihre ultimately this is your decision, I am just talking through other options because the js bundles that are produced are pretty gnarly, and seeing them in all UI commits will get ugly

**Reply by Safihre** [reply]
Let's first get the basic (no translations or compiled files in repo) running and take it from there :)

**Reply by sbalbrecht** [reply]
Hi, quick update from me. I've been working steadily on a [branch](https://github.com/sbalbrecht/sabnzbd/tree/feature/react_ui) over the last couple weeks with initial setup. I migrated the first page of the wizard as a proof of concept, which currently works identically to how it did before. I ended up loading the same bootstrap/jquery libs directly in react to maintain styling and will deal with the consequences of that once I reach the first page that actually uses jquery. I have not yet worked out the kinks for sab serving the react UI alongside the existing one, but running the node dev server you can access the whole app.

I also had to add/tweak a couple of endpoints in order to allow the wizard to fetch the header, list of languages and the translation keys. This was just to get myself up and running, Salfihre can change them if necessary.

Localization is the last thing I want to figure out before starting to migrate other pages, because it will be a pain to have to change all the localized strings to a different syntax later, and I want to get loading translations dynamically out of the way.

I'm looking for a lib that will allow us to continue using the po files in the project. i18next is popular but uses json for translations. I have been experimenting with LinguiJS as it works with po files. Under normal operation it generates po files from your UI code during compilation, and then loads the po file for the current language in the UI. This seemed promising, but it failed when I tried to instead load one of the existing po files, because apparently behind the scenes they convert the contents of their generated po into a json object, even though the extension is still ".po".

Here’s an example from one of their [example apps](https://github.com/lingui/js-lingui/tree/main/examples/vite-project-react-babel):

`pl.po` generated by Lingui on the filesystem:

```po
msgid ""
msgstr ""
"POT-Creation-Date: 2023-03-23 14:43+0100\n"
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=utf-8\n"
"Content-Transfer-Encoding: 8bit\n"
"X-Generator: @lingui/cli\n"
"Language: pl\n"
"Project-Id-Version: \n"
"Report-Msgid-Bugs-To: \n"
"PO-Revision-Date: \n"
"Last-Translator: \n"
"Language-Team: \n"
"Plural-Forms: \n"

#: src/App.tsx:35
msgid "{count, plural, one {# month} other {# months}}"
msgstr "{count, plural, one {# miesiąc} few {# miesiące} many {# miesięcy} other {# miesiąca}}"

#: src/App.tsx:53
msgid "Click on the Vite and React logos to learn more"
msgstr "Kliknij na logo Vite i React, aby dowiedzieć się więcej"

#: src/App.tsx:38
msgid "Edit <0>src/App.tsx</0> and save to test HMR"
msgstr "Edytuj <0>src/App.tsx</0> i zapisz, aby przetestować HMR"
```

`pl.po` in the browser network tab when loading the page:

```po
/*eslint-disable*/
export const messages = JSON.parse("{\"XmK1ih\":[[\"count\",\"plural\",{\"one\":[\"#\",\" miesiąc\"],\"few\":[\"#\",\" miesiące\"],\"many\":[\"#\",\" miesięcy\"],\"other\":[\"#\",\" miesiąca\"]}]],\"UNWX+z\":\"Kliknij na logo Vite i React, aby dowiedzieć się więcej\",\"JAd81L\":\"Edytuj <0>src/App.tsx</0> i zapisz, aby przetestować HMR\"}");
```

Not really expecting help with a solution, just wanted to show the odd implementation. Cheers!

**Reply by Safihre** [reply]
Awesome. That po file looks the same like ours, right?

**Reply by sbalbrecht** [reply]
Yep!

**Reply by thezoggy** [reply]
is there any good way to measure the impact/benefit of the front end stack change in a meaningful way?

with sab paused (so we focus on front end and not adding the backend downloading stuff), refreshing the front page.
how much memory used / how long it took to load / cpu usage... when sab is idle with empty queue vs lets say 10k items in the queue.

i know things would be imperfect but ballpark to know hey we know the change is going to yield net benefits to warrant the change.

**Reply by sbalbrecht** [reply]
Yeah, we can measure bundle sizes for each page load before and after, both initial page load and load when navigating, and initial rendering time of the content. I expect the SPA to be snappier and lighter during navigation, and the initial bundle might even be smaller too. I can take some of those measurements when I'm a bit further along, but yeah there are a bunch of metrics to compare.

**Reply by Safihre** [reply]
@thezoggy but I doubt the load on the server side will be lower. I already squeezed out as much as I could from the main API calls.
Replacing Cherrypy will help I think, but that's a good one to measure for huge queues.

**Comment by sbalbrecht** [comment]
Hey all, checking in with a hefty update.

Following up from last week, I figured out translation with LinguiJS and now translation in the UI is as simple as using the `<Trans>` tag or `t`  macro using the English string directly, which I personally find much more readable.

```tsx
<Trans>SABnzbd Quick-Start Wizard</Trans>
```

```tsx
<input title={t`Select only if your provider allows SSL connections.`}
    ... other attributes ...
/>
```

I spent this week migrating `/wizard/one`. It was a good stepping stone as it required me to get familiar with libraries for form state management, routing, and data fetching. I also migrated the [imperative state management logic](https://github.com/sbalbrecht/sabnzbd/blob/feature/react_ui/interfaces/wizard/static/javascript/checkserver.js#L15-L32) to [React’s declarative approach](https://github.com/sbalbrecht/sabnzbd/blob/feature/react_ui/ui/src/routes/wizard/one.tsx#L331-L370), which was satisfying. The only jQuery I haven’t yet migrated on the page are the tooltips.

There are a few improvements to make regarding routing and data fetching on each page load due to my inexperience with the libraries and patterns, but overall I think this is a good foundation for how most functionality can be migrated.

However, now that I have a better idea of what the migration will look like, and before I go much farther, I want to raise some concerns about the decision to introduce a node-based frontend framework to the project.

I recently reread [https://grugbrain.dev/](https://grugbrain.dev/) (among other essays) and its warnings of the dreaded complexity demon. The [section on front-end development](https://grugbrain.dev/#grug-on-front-end-development) struck a chord:

> "I know, I'll split my front end and back end codebase up and use a hot new SPA library talking to a GraphQL JSON API back end over HTTP (which is funny because I'm not transferring hypertext)"
>
>now you have two complexity demon spirit lairs
>
>and, what is worse, front end complexity demon spirit even more powerful and have deep spiritual hold on entire front end industry as far as grug can tell 

The existing UI is kinda fragmented with different outdated libraries, but the stack is super simple with Python serving the Cheetah templates.

With React introduced, there are already a lot of moving pieces:

- node.js and npm as a separate runtime during development and compilation
- [React](https://react.dev/) as the core framework for state management and UI componentization
- [Vite](https://vitejs.dev/) for the dev server and bundler
- [LinguiJS](https://lingui.dev/) for localization
- [@tanstack/query](https://tanstack.com/query/latest/docs/framework/react/overview) for data fetching
- [@tanstack/router](https://tanstack.com/router/latest/docs/framework/react/overview) for routing (could instead go with react-router, but I liked the type safety and unification with query)
- [react-hook-form](https://react-hook-form.com/get-started) for form state management
- [zod](https://zod.dev/) for form schema validation to be used with react-hook-form
- Bootstrap still remains, but could be replaced with something like tailwindcss

Even though I haven’t yet explored the requirements of Glitter and the Config, I don’t anticipate too many more big libs for the frontend, but the point stands.

After some research, it seems like Cheetah, Knockout.js and jQuery can be replaced with more modern alternatives without devoting ourselves to the heaviness of a JavaScript frontend framework.

For example:

- Starlette docs recommend [Jinja](https://jinja.palletsprojects.com/en/3.1.x/) for templating
- Embracing templating over JavaScript, Knockout.js and jQuery could be replaced with [htmx](https://htmx.org/docs/) for HATEOAS, using something like [alpinejs](https://alpinejs.dev/) to supplement the scripty-bits that are hard to implement using hypertext alone.
- Bootstrap stays for now but could be upgraded to the latest version that drops jQuery.

While this is a less conventional stack given the state of frontend today, it is **simple** (and I mean htmx was 2nd in the [2023 JavaScript Rising Stars](https://risingstars.js.org/2023/en#section-framework) "Front-end Frameworks" category, [#10 overall](https://risingstars.js.org/2023/en#section-all)).

You get to keep the benefits and speed of SSR, have relatively few dependencies, and keep everything within the Python backend instead of splitting the frontend.

I get the draw to use a JavaScript SPA framework because they are the predominant solution today and there are plenty of resources, but that decision adds a lot of complexity to a previously Python-exclusive project. As we have established that there are few contributors to SABnzbd, so the most important thing is that the stack is easy to digest for existing maintainers. Making the UI in React is doable, but maybe more complicated than SABnzbd really requires.

Let me know your thoughts. I would be willing to start a new branch to see what the Jinja/htmx/alpine stack might look like.

**Reply by thezoggy** [reply]
I picked this up last year, been a pretty good read:
https://livebook.manning.com/book/microservice-apis-in-python/

at work we use jinja2 and its not bad.  of course there there deff some fundamental issues that can arise if you dont know up front if you want to do logic in the templates or just do it all outside of it. but prob true of most template engines.

**Reply by Safihre** [reply]
Good that you brought up the point, it was brought up by developers at my work too with whom I discussed SABnzbd front-end. It's quite heavy for a single-page application..
For Glitter we do need quite some Javascript to make the update to the data from the API results. I don't think we can do with just the 3 items you discussed. 

So what do we do? Is there a middle ground? Or is it really jQuery/Knockout on one side and heavy-front-end on the other?

**Reply by sbalbrecht** [reply]
I will take a closer look at the Glitter js, but I do think all the interactivity is achievable with vanilla js and htmx/alpine with some ingenuity. We can eliminate knockout and jquery without moving to a js SPA framework. I'll give an update in a week with some thoughts.

**Reply by sbalbrecht** [reply]
> I picked this up last year, been a pretty good read:
> 
> https://livebook.manning.com/book/microservice-apis-in-python/
> 
Also, I will take a look!

**Reply by sbalbrecht** [reply]
Hi all. Life got busy for me over the last two weeks and isn't stopping, so I haven't looked into this migration directly since my last update.

However, I had the opportunity to rewrite a UI at work using thymeleaf (Java templating engine) and htmx and learned a lot.

I was able to implement infinite scrolling, polling server status, replacing table rows, etc., without any JavaScript. The only thing I used JS for was for sorting table columns, but this was a preexisting solution that I didn't have the time to migrate -- still doable.

With this method of development, instead of receiving JSON from the server and JS updating the UI based on its state, you receive a block of HTML with which you replace a portion of the UI, without refreshing the entire page.

For example, if you want to [edit a table row](https://htmx.org/examples/click-to-edit/), you click the "Edit" button on that row, which sends an HTTP GET, and the response is that same row with inputs inside the table cells. After editing you click "Save", which sends an HTTP PUT, and the response is the updated row with static content. The entire state of the UI is communicated with hypermedia alone, without JS.

This means that most of the logic for configuring the UI according to the state of the data is handled by the templating engine at render time, using endpoints to handle changes in UI state instead of handling changes with client-side code.

You are potentially creating more back and forth between client and server to handle UI changes, but you are also aren't shipping JS bundles which would normally be used to do that work. Since most apps already contact the server to communicate state updates with JSON, this seems to balance out.

Another consideration is that HTML is a heavier format that JSON, but again, the JSON requires JS to interpret so the UI can be updated, so it balances out in a similar way.

*Also worth noting htmx recommends splitting your application API from your general purpose data API as opposed to using content negotiation in the same endpoint, mainly because the general API can be kept stable while the app API changes with features (and, only server hypermedia).*

So I have found communicating state with hypermedia works well for a lot of seemingly complex interactions, but becomes more difficult to work with if the app has many dynamic components, i.e., you will have a bad time implementing Google Sheets using hypermedia alone.

Glancing through the Glitter JS, I suspect a lot of the Knockout functionality can be migrated to this approach, but y'all would know better than me about tricky parts. I still plan on trying it out when I have more time.

**Reply by Safihre** [reply]
You describe exactly how SABnzbd used to work 10 years ago 😅😅 all the skins worked exactly like that, using the template logic. 
It was an utter mess...

**Reply by sbalbrecht** [reply]
Oof!

I sense you do not want to go back down that route. In the meantime, migrating from Cheetah to Jinja shouldn't be too bad.
